[WIP][serve] Experimental support for websockets

edoakes opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

edoakes commented on python/ray/serve/_private/http_util.py:
```diff
@@ -144,9 +139,10 @@ def status_code(self):
-class ASGIHTTPSender(Send):
-    """Implement the interface for ASGI sender to save data from varisous
-    asgi response type (fastapi, starlette, etc.)
+class BufferedASGISender(Send):
```
This is just a rename (the previous name was very generic)

edoakes replied:
split into separate PR:
https://github.com/ray-project/ray/pull/36168

GeneDer commented on python/ray/serve/_private/replica.py:
```diff
@@ -494,95 +515,74 @@ def callable_method_filter(attr):
+            if isinstance(self.callable, ASGIAppReplicaWrapper):
+                request_args = (scope, receive, send)
+            else:
+                request_args = (Request(scope, receive, send),)
```
Non-blocker: The way this is packed and unpacked, maybe we can just do 
```
if not isinstance(self.callable, ASGIAppReplicaWrapper):
    request_args = (Request(*request_args),)
```

shrekris-anyscale commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -75,10 +75,12 @@
     scope: Scope,
     receive: Receive,
     send: Send,
+    request_id: str,
```
Could you remove the `request_id` argument? It isn't used in this function.

GeneDer commented on python/ray/serve/_private/replica.py:
```diff
@@ -494,95 +515,74 @@ def callable_method_filter(attr):
                 # ASGI interace and result should always be `None`.
-                if callable_is_asgi_wrapper:
+                if isinstance(self.callable, ASGIAppReplicaWrapper):
                     assert result is None
```
Non-blocker: Would be nice if we can also log a message explain this assertion. But knowing this was already like that before and can be improved in a followup.

shrekris-anyscale commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -106,14 +108,9 @@ async def _handle_streaming_response(
-                        "'http.response.start' message must contain 'status'",
-                    )
-                    status_code = str(asgi_message["status"])
+                    status_code = str(asgi_message.get("status", "UNKNOWN"))
```
Why don't we raise an `AssertionError` based on `type` or `status` anymore?

edoakes replied:
because we now support websockets which have a different type and do not have a status

edoakes replied:
updated the comment and explicitly checking message type now

GeneDer commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -106,14 +108,9 @@ async def _handle_streaming_response(
-                        "'http.response.start' message must contain 'status'",
-                    )
-                    status_code = str(asgi_message["status"])
+                    status_code = str(asgi_message.get("status", "UNKNOWN"))
```
Nitpick: Not sure how often this `UNKNOWN` status will be used. Maybe we can refactor a constant for reusing

edoakes replied:
fixed so we don't have it anymore :)

shrekris-anyscale commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -490,24 +394,34 @@ async def __call__(self, scope, receive, send):
-        status_code = await _send_request_to_handle(handle, scope, receive, send)
+
+        if RAY_SERVE_ENABLE_EXPERIMENTAL_STREAMING:
+            status_code = await self.send_request_to_replica_streaming(
```
[Nit] Could you rename the following functions:

* `send_request_to_replica_streaming` -> `send_streaming_request_to_replica`
* `send_request_to_replica_unary` -> `send_unary_request_to_replica`

This way, the adjective is closer to the noun.

edoakes replied:
It's not actually a "streaming request" vs. a "unary request," but it's the method of sending the same type of request... but that might have made it even more confusing ;)

A longer form way of saying this would be "send_request_to_replica_using_unary_protocol" or "send_request_to_replica_using_streaming_protocol"

shrekris-anyscale replied:
Oh that makes sense, thanks for explaining!

shrekris-anyscale commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -535,19 +449,191 @@ async def __call__(self, scope, receive, send):
                 }
             )
 
+    async def send_request_to_replica_unary(
```
To confirm, this code is the exact same as `_send_request_to_handle`, correct?

edoakes replied:
yup this is unchanged

shrekris-anyscale commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -535,19 +449,191 @@ async def __call__(self, scope, receive, send):
+            return "200"
+
+    async def proxy_asgi_receive(self, receive: Receive, queue: asyncio.Queue):
+        """TODO"""
```
Could you elaborate or remove the `TODO`?

edoakes replied:
done

GeneDer commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -535,19 +449,191 @@ async def __call__(self, scope, receive, send):
+            return "200"
+
+    async def proxy_asgi_receive(self, receive: Receive, queue: asyncio.Queue):
+        """TODO"""
```
Maybe complete the docstring 🙃

shrekris-anyscale commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -535,19 +449,191 @@ async def __call__(self, scope, receive, send):
+
+            # If the first message has been sent we cannot send a 500 status as it
+            # would violate ASGI protocol.
+            # TODO(edoakes): this may not be valid at all for WebSocket requests.
```
Should we add a test for this?

edoakes replied:
removed this logic entirely. we'll still return an `Internal Server Error` in this case

GeneDer commented on python/ray/serve/tests/test_fastapi.py:
```diff
@@ -118,13 +118,23 @@ def other(self, msg: str):
 
 
-def test_make_fastapi_cbv_util():
+@pytest.mark.parametrize("websocket", [False, True])
```
Great way to test it!

GeneDer reviewed (commented):
Some nitpick before merging, but generally lgtm :)

shrekris-anyscale commented on python/ray/serve/_private/http_util.py:
```diff
@@ -172,7 +168,9 @@ def __init__(self):
-        assert message["type"] in ("http.response.start", "http.response.body")
+        # assert message["type"] in ("http.response.start", "http.response.body")
+        if "websocket" in message["type"]:
+            print("PUT WEBSOCKET MESSAGE", message)
```
Change `print` to `logging`.

shrekris-anyscale commented on python/ray/serve/_private/http_util.py:
```diff
@@ -172,7 +168,9 @@ def __init__(self):
 
     async def __call__(self, message: Dict[str, Any]):
-        assert message["type"] in ("http.response.start", "http.response.body")
+        # assert message["type"] in ("http.response.start", "http.response.body")
```
Remove commented code.

shrekris-anyscale commented on python/ray/serve/_private/replica.py:
```diff
@@ -494,95 +515,74 @@ def callable_method_filter(attr):
 
-    async def send_user_result_over_asgi(self, result: Any, asgi_sender: Send):
+    async def send_user_result_over_asgi(
+        self, result: Any, scope: Scope, asgi_sender: Send, asgi_receive: Receive
```
[Nit] The code uses both `asgi_sender` (noun) and `asgi_receive` (verb). Could we pick one tense and keep it consistent?

edoakes replied:
yes I'm going to rename these to asgi standard names everywhere (just `send` and `receive`)

edoakes replied:
updated

shrekris-anyscale reviewed (commented):
Nice work so far! I left some comments.

edoakes commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -106,14 +108,9 @@ async def _handle_streaming_response(
             for asgi_message in asgi_messages:
                 # There must be exactly one "http.response.start" message that
                 # always contains the "status" field.
+                # TODO: fix.
```
updated

sihanwang41 reviewed (commented):
High level lgtm!
Worth adding test scenario:
1. websocket & StreamingResponse in the same Deployment Class.
2. HTTP Proxy dead during the stream. (Make sure replica can handle the exception since we passes the handle)

edoakes commented:
@sihanwang41:

- I added a test to `http_util` to verify that `receive` raises if the actor crashes or raises an exception.
- Manually tested that a generator/websocket exits if HTTP proxy crashes.
- Added a test case for one deployment with a normal HTTP response, streaming HTTP response, and websocket.

edoakes commented:
Also added a test that `request.is_disconnected()` now works as expected.

shrekris-anyscale commented on python/ray/serve/_private/http_util.py:
```diff
@@ -156,8 +158,8 @@ def build_asgi_response(self) -> RawASGIResponse:
-class ASGIHTTPQueueSender(Send):
-    """ASGI sender that enables polling for the sent messages off a queue.
+class ASGIMessageQueue(Send):
+    """Queue enables polling for received or sent messages off a queue.
```
```suggestion
    """Queue enables polling for received or sent messages.
```

GeneDer commented on python/ray/serve/_private/replica.py:
```diff
@@ -218,78 +213,113 @@ async def handle_request(
+
+            request_metadata = pickle.loads(pickled_request_metadata)
+            if request_metadata.is_http_request:
+                assert len(request_args) == 1
```
Non-blocker: Would be nice to add a message here to note what this is checking. HTTP request requires exactly one request argument or something similar :)

shrekris-anyscale reviewed (approved):
Nice work, this change looks good to me!

GeneDer commented on python/ray/serve/_private/replica.py:
```diff
@@ -494,95 +515,74 @@ def callable_method_filter(attr):
 
-        args, kwargs = parse_request_item(request_item)
+        if request_metadata.is_http_request:
+            assert len(request_args) == 3
```
And maybe add a message if possible, but not blocker :)

GeneDer reviewed (approved):
LGTM!

edoakes commented on python/ray/serve/_private/http_util.py:
```diff
@@ -195,6 +196,75 @@ async def wait_for_message(self):
+                )
+                for message in pickle.loads(pickled_messages):
+                    if message["type"] in {"http.disconnect", "websocket.disconnect"}:
+                        self._disconnect_message = message
```
I'll add a comment; it's cancelled by the caller
